include/onnxruntime/core/common/gpu_profiler_common.h

```diff
@@ -0,0 +1,397 @@
+// is compiled. Thus, the base class symbols will not be included in the
+// main onnxruntime shared library. This manifests in being unable to load
+// EP shared libs (because the base class symbols referenced by derived
+// classes are missing).
+// We solve this by implementing base classes that are common to all GPU profilers
+// inline in this header.
+
+class ProfilerActivityBuffer {
+ public:
+  ProfilerActivityBuffer()
+      : data_(nullptr), size_(0) {}
+
+  ProfilerActivityBuffer(const char* data, size_t size)
+      : data_(std::make_unique<char[]>(size)), size_(size) {
+    memcpy(data_.get(), data, size_);
+  }
+
+  ProfilerActivityBuffer(const ProfilerActivityBuffer& other)
+      : ProfilerActivityBuffer(other.GetData(), other.GetSize()) {}
+
+  ProfilerActivityBuffer(ProfilerActivityBuffer&& other)
+      : ProfilerActivityBuffer() {
+    std::swap(data_, other.data_);
+    std::swap(size_, other.size_);
+  }
+
+  ProfilerActivityBuffer& operator=(const ProfilerActivityBuffer& other) {
+    if (&other == this) {
+      return *this;
+    }
+
+    new (this) ProfilerActivityBuffer{other};
+    return *this;
+  }
+
+  ProfilerActivityBuffer& operator=(ProfilerActivityBuffer&& other) {
+    if (&other == this) {
+      return *this;
+    }
+
+    new (this) ProfilerActivityBuffer{std::move(other)};
+    return *this;
+  }
+
+  static ProfilerActivityBuffer CreateFromPreallocatedBuffer(char* data, size_t size) {
+    ProfilerActivityBuffer res{};
+    res.data_.reset(data);
+    res.size_ = size;
+    return res;
+  }
+
+  // accessors
+  char* GetData() { return data_.get(); }
+  const char* GetData() const { return data_.get(); }
+  size_t GetSize() const { return size_; }
+
+ private:
+  std::unique_ptr<char[]> data_;
+  size_t size_;
+}; /* end class ProfilerActivityBuffer */
+
+template <typename TDerived>
+class GPUTracerManager {
+ public:
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(GPUTracerManager);
+  virtual ~GPUTracerManager() {}
+
+  uint64_t RegisterClient() {
+    std::lock_guard<std::mutex> lock(manager_instance_mutex_);
+    auto res = next_client_id_++;
+    per_client_events_by_ext_correlation_.insert({res, {}});
+    ++num_active_clients_;
+    return res;
+  }
+
+  void DeregisterClient(uint64_t client_handle) {
+    std::lock_guard<std::mutex> lock(manager_instance_mutex_);
+    auto it = per_client_events_by_ext_correlation_.find(client_handle);
+    if (it == per_client_events_by_ext_correlation_.end()) {
+      return;
+    }
+    per_client_events_by_ext_correlation_.erase(it);
+    --num_active_clients_;
+    if (num_active_clients_ == 0 && logging_enabled_) {
+      StopLogging();
+    }
+  }
+
+  void StartLogging() {
+    std::lock_guard<std::mutex> lock(manager_instance_mutex_);
+    if (logging_enabled_) {
+      return;
+    }
+
+    auto this_as_derived = static_cast<TDerived*>(this);
+    logging_enabled_ = this_as_derived->OnStartLogging();
+  }
```

Comment:
The `logging` here seems to be a little bit easy to be confused with a Logger. Maybe use `tracing` or `profiling` shall be used instead.

Reply:
Better using `tracing`, because `GPUTracerManager` manages `Tracer`s and `Tracer`s will do the tracing.

Reply:
Good idea, thanks! Changed to `tracing_enabled_`.